res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+}
+
+TraktorS2MK2.scalerParameter = function(group, name, value) {
+  return script.absoluteLin(value, 0, 1, 16, 4080);
+}
+// Tell the HIDController script to use setParameter instead of setValue.
+TraktorS2MK2.scalerParameter.useSetParameter = true;
+
+TraktorS2MK2.scalerVolume = function(group, name, value) {
+  if (group === "[Master]") {
+    return script.absoluteNonLin(value, 0, 1, 4, 16, 4080);
+  } else {
+    return script.absoluteNonLin(value, 0, 0.25, 1, 16, 4080);
+  }
+}
+
+TraktorS2MK2.scalerSlider = function(group, name, value) {
+  return script.absoluteLin(value, -1, 1, 16, 4080);
+}
+
+TraktorS2MK2.outputChannelCallback = function(value,group,key) {
+  var led_value = 0x05;
+  if (value) {
+    led_value = 0x7F;
+  }
+  TraktorS2MK2.controller.setOutput(group, key, led_value, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.outputChannelCallbackDark = function(value,group,key) {
+  var led_value = 0x00;
+  if (value) {
+    led_value = 0x7F;
+  }
+  TraktorS2MK2.controller.setOutput(group, key, led_value, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.outputCallback = function(value,group,key) {
+  var led_value = 0x09;
+  if (value) {
+    led_value = 0x7F;
+  }
+  TraktorS2MK2.controller.setOutput(group, key, led_value, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.outputCallbackLoop = function(value,group,key) {
+  var led_value = 0x09;
+  if (engine.getValue(group, "loop_enabled")) {
+    led_value = 0x7F;
+  }
+  TraktorS2MK2.controller.setOutput(group, key, led_value, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.outputCallbackDark = function(value,group,key) {
+  var led_value = 0x00;
+  if (value) {
+    led_value = 0x7F;
+  }
+  TraktorS2MK2.controller.setOutput(group, key, led_value, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.pflButton = function(field) {
+  if (field.value > 0) {
+    if (TraktorS2MK2.shift_pressed[field.group]) {
+      script.toggleControl(field.group, "quantize");
+    } else {
+      script.toggleControl(field.group, "pfl");
+    }
+  }
+}
+
+TraktorS2MK2.sendPadColor = function(group, padNumber, color) {
+  var padKey = "!pad_" + padNumber + "_";
+  TraktorS2MK2.controller.setOutput(group, padKey + "R", color.red, false);
+  TraktorS2MK2.controller.setOutput(group, padKey + "G", color.green, false);
+  TraktorS2MK2.controller.setOutput(group, padKey + "B", color.blue, !TraktorS2MK2.freeze_lights);
+}
+
+TraktorS2MK2.outputHotcueCallback = function(value, group, key) {
+  var hotcueNumber = key.charAt(7);
+  var color;
+  if (engine.getValue(group, 'hotcue_' + hotcueNumber + '_enabled')) {
+    color = colorCodeToObject(engine.getValue(group, 'hotcue_' + hotcueNumber + '_color'));
+  } else {
+    color = {red: 0, green: 0, blue: 0};
+  }
+  TraktorS2MK2.sendPadColor(group, hotcueNumber, color);
+}
+
+TraktorS2MK2.onVuMeterChanged = function(value, group, key) {
+  // This handler is called a lot so it should be as fast as possible.
```

Comment:
TODO: we should also make this generic, vu meters are super common

Reply:
How could this be generic? Every controller implements communication with meter LEDs differently.

Reply:
function(volume, num_segments) {
return [an array of values 0-255]
}